File: 91-decode-ways/91-decode-ways.py
class Solution:
    def numDecodings(self, s: str) -> int:
        n = len(s)
        dp = [0]*n
        flag = True
        if s[0] == '0':
            return 0
        for ind, char in enumerate(s):
            if char != '0' and ind == 0:
                dp[ind] = 1
            elif char == '0':
                if s[ind-1] == '1' or s[ind-1] == '2':
                    if ind > 1:
                        dp[ind] = dp[ind-2]
                    else:
                        dp[ind] = 1
                else:
                    dp[ind] = 0
            elif char != '0' and ind != 0:
                if ind > 1:
                    if (s[ind-1] == '2' and s[ind] < '7') or (s[ind-1] == '1'):
                        dp[ind] = dp[ind-1] + dp[ind-2]
                    else:
                        dp[ind] = dp[ind-1]
                else:
                    if (s[ind-1] == '2' and s[ind] < '7') or (s[ind-1] == '1'):
                        dp[ind] = dp[ind-1] * 2
                    else:
                        dp[ind] = dp[ind-1]
        return dp[-1]
                    

File: 91-decode-ways/91-decode-ways.java
class Solution {
    
    public int numDecodings(String s) {
        //@input: string
        //@output : number of ways to decode
        //@edge case : leading 0 
        //@breaking down problem:
        // 1. set the recursion funciton with String s, curRes as input, the final res as output. 
        // 2. in each recursion step, select the first or the first two numbers as the current choice.
        // 3. then the total number of ways is equal to the sum of returned values from the two sub-functions 
        // 4. return the total number of ways as the final result
        //5. when the string is empty, terminate the recursion
        int n = s.length();
        int oneBack = 1, twoBack = 0;
        int temp = 0;
        for(int i = 1; i < n + 1; i++){
            int firstNum = Integer.parseInt(s.substring(i-1, i));
            temp = 0;
            if(firstNum != 0){
                temp += oneBack;
            }
            if (i > 1){
                int secondNum = Integer.parseInt(s.substring(i-2, i - 1));
                int totalNum = secondNum * 10 + firstNum;
                if(totalNum <= 26 && secondNum > 0){
                    temp += twoBack;
                }
            }
            twoBack = oneBack;
            oneBack = temp;
            
            
            
        }
        return temp;
    }
    


    
}